cpp/include/cudf/detail/replace/nulls.cuh

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2018-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/types.hpp>
+
+#include <thrust/functional.h>
+
+namespace cudf {
+namespace detail {
+
+using idx_valid_pair_t = thrust::tuple<cudf::size_type, bool>;
+
+/**
+ * @brief Functor used by `replace_nulls(replace_policy)` to determine the index to gather from in
+ * the result column.
+ *
+ * Binary functor passed to `inclusive_scan` or `inclusive_scan_by_key`. Both parameter and return
+ * type are a tuple of cudf::size_type and boolean. For parameter, the first of the tuple is the
+ * index of the row, and the second is the validity of that row. For return value, the first is the
+ * index in the gather map for the output column. If the current row in input is NULL, it is the
+ * index accumulated along the iteration direction; if the current row in input is valid, it is the
+ * index of current row. The second element of the return tuple is discarded.
+ */
+struct replace_policy_functor {
+  __device__ idx_valid_pair_t operator()(idx_valid_pair_t const& lhs, idx_valid_pair_t const& rhs);
```

Comment:
Why is the definition of this function somewhere else?

Reply:
Offline discussion: device code cannot have separated function definition and declaration unless configured so: https://developer.nvidia.com/blog/separate-compilation-linking-cuda-device-code/

Will move definition to `nulls.cuh`.